---
output: github_document
---

<!-- README.md is generated from README.Rmd. Please edit that file -->

```{r, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>",
  fig.path = "figures/README-",
  out.width = "100%"
)
```

## DeepG4ToolsComparison: A snakemake pipeline to run and compare G4 DNA prediction tools with DeepG4

This pipeline generates the `precision/recall` curves on genome-wide using `201 bp` bins. It also generate the `precision/recall` curves on [promoters](bed/Genes_hg19_from_Homo.sapiens.bed).


### Overview


It's based on [Snakemake](https://snakemake.readthedocs.io/en/stable/) to manage the workflow and [Docker](https://www.docker.com/) to isolate the application and run it with the appropriate tool versions.


### Installation

#### Clone the repository :

```bash
git clone https://github.com/morphos30/DeepG4ToolsComparison.git
cd DeepG4ToolsComparison
```
#### Install the docker image and run it :

```bash
docker build . -t morphos30/g4docker -f Dockerfile/Dockerfile
docker run -it -v $(pwd):/DeepG4ToolsComparison morphos30/g4docker /bin/bash
```

Where `$(pwd)` is the working directory of `DeepG4ToolsComparison` on your computer.


#### Launch the pipeline :

```bash
cd /DeepG4ToolsComparison/prediction_genome
snakemake --use-conda
```

You have to set the option `--use-conda` in order to install and run each tool in its proper environment.

### Workflow specifications

Unfortunately, bigwig files are too big to be stored on github, so the user will have to download them in order to launch the pipeline. Informations to get the files are available in the [bigwig directory](bigwig/).

#### Input (auPR)

* DNA sequences into bed format written into the bed directory.

__Note :__ if you want add a new dataset, edit the `Snakefile` file and add the bed files in the dictionnary `EXPERIMENTS`, without the `.bed` extension. Example :

`TestSet_Peaks_BG4_G4seq_HaCaT_GSE76688_hg19_201b_Ctrl_gkmSVM_0.8_42.bed`

```python
EXPERIMENTS = {
  "TestSet_Peaks_BG4_G4seq_HaCaT_GSE76688_hg19_201b_Ctrl_gkmSVM_0.8_42_Ctrl_gkmSVM":{"EXP":"TestSet_Peaks_BG4_G4seq_HaCaT_GSE76688_hg19_201b_Ctrl_gkmSVM_0.8_42"}
}
```

Where `CTRL` is the negative set and `EXP` is the positive set.



* DNA Accessibility (ATAC-seq/DNAse-seq/MNase-seq) in bigwig format, stored in the `bigwig` directory.

```python
ATACFILE = {
	"TestSet_Peaks_BG4_G4seq_HaCaT_GSE76688_hg19_201b_Ctrl_gkmSVM":["ATAC_entinostat_mean"]
}
```

#### Input (Prediction only)

* DNA Accessibility (ATAC-seq/DNAse-seq/MNase-seq) in bigwig format, stored in the `bigwig` directory.

```python
TISSUES = [
"cerebellum_tissue"
]

ATACFILE = {
	"cerebellum_tissue":["ENCODE/cerebellum_tissue"]
}
```

#### Rulegraph :

```{r, echo=FALSE, out.width="150%", fig.cap=""}
knitr::include_graphics("rulegraph.svg")
```

### Main result

```{r boxplot1, fig.height=6, fig.width=10,dev="png",dpi=150,echo=F,warning=F,message=F}
require(tidyverse)
require(glue)

theme_set(theme_classic(base_size=18))
theme_update(legend.text=element_text(size=12),
             legend.direction = "vertical",
             # legend.box = "horizontal",
             # legend.position = c(1,0.025),
             legend.position = "bottom",
             axis.title.y = element_blank(),
             axis.ticks.x = element_blank(),
             axis.text.x = element_blank(),
             legend.justification = c(0, 0),
             axis.text.y = element_text(color = "black",size=16))

AUPR_data <- c(list.files("results/AUC/",pattern=".tsv$",full.names=T)) %>% setNames(basename(.)) %>% map(read_tsv) %>% bind_rows(.id="file") %>% 
  filter(!str_detect(file,"qG4-ChIP-seq-of-breast-cancer-PDTX")) %>% 
  filter(!str_detect(file,"DENOISED_"))

AUPR_data <- AUPR_data %>% 
  mutate(Exp = str_extract(file,"GSE[0-9]+|breast-cancer-PDTX"))  %>% 
  mutate(Tool = str_extract(basename(file),"\\..+")) %>%
  mutate(Tool = str_remove_all(Tool,"\\.|tsv$")) %>%
  mutate(cell_line = str_extract(file,"qG4|HaCaT|K562|HEKnp|H1975|A549|293T|HeLaS3")) %>% 
  mutate(TypeExp = str_extract(file,"Peaks_G4P_G4seq|TestSet_Peaks_qG4|TestSet_Peaks_G4seqpm_BG4|TestSet_Peaks_BG4_G4seq|Promoters_qG4|Promoters_G4seq_BG4|Promoters_BG4_G4seq|Peaks_G4seqpm_BG4|Peaks_BG4_G4seq|Peaks_G4seq_qG4|Peaks_G4seq_BG4|Peaks_qG4|Peaks_G4seqpm_qG4")) 
##Classement

AUPR_data <- AUPR_data %>%
  mutate(x_axis = glue("**{cell_line}** *({Exp})*")) %>% 
  mutate(x_order = str_c(Exp,cell_line,sep="_")) %>% 
  mutate(x_order = as.numeric(as.character(fct_recode(as.factor(x_order),`1`="GSE76688_HaCaT",`2`="GSE99205_HaCaT",`3`="GSE76688_HEKnp",`4`="GSE107690_K562",`5`="GSE133379_HeLaS3",`6`="GSE133379_293T",`7`="GSE133379_A549",`8`="GSE133379_H1975")))) %>% 
  # filter(Tool %in% c("DeepG4ATACnormBG","DeepG4_classictuningOH5","G4detector_tsv","G4detector_retrained_tsv","penguinn_tsv","penguinn_retrained_tsv","quadron_retrained","quadron_score")) %>%
  filter(Tool %in% c("ATACDeepG4_ATACnormBG","DeepG4_classictuningOH5","G4detector_tsv","penguinn_tsv","quadron_score")) %>%
  # mutate(Tool_color = fct_recode(as.factor(Tool),`#e84118`="ATACDeepG4_ATACtuningOH5",`#273c75`="penguinn_retrained_tsv",`#487eb0`="penguinn_tsv",`#353b48`="G4detector_retrained_tsv",`#7f8fa6`="G4detector_tsv")) %>% 
  # mutate(Tool = fct_recode(as.factor(Tool),`DeepG4`="ATACDeepG4_ATACtuningOH5",`DeepG4*`="DeepG4_classictuningOH5",`DeepG4\n(normBG)`="DeepG4ATACnormBG",`Penguinn retrained`="penguinn_retrained_tsv",`Penguinn`="penguinn_tsv",`G4detector retrained`="G4detector_retrained_tsv",G4detector="G4detector_tsv")) %>% 
  # mutate(Tool = fct_recode(as.factor(Tool),`DeepG4*`="DeepG4_classictuningOH5",`DeepG4`="DeepG4ATACnormBG",`Penguinn retrained`="penguinn_retrained_tsv",`Penguinn`="penguinn_tsv",`G4detector retrained`="G4detector_retrained_tsv",G4detector="G4detector_tsv",`Quadron retrained`="quadron_retrained",Quadron="quadron_score")) %>% 
  mutate(Tool = fct_recode(as.factor(Tool),`DeepG4*`="DeepG4_classictuningOH5",`DeepG4`="ATACDeepG4_ATACnormBG",`Penguinn`="penguinn_tsv",G4detector="G4detector_tsv",Quadron="quadron_score")) %>% 
  # mutate(Tool = glue("<i style='color:{Tool_color}'><b>{Tool}</b></i>")) %>% 
  group_by(cell_line,Exp) %>% 
  arrange(desc(AUC)) %>% mutate(Classement = 1:dplyr::n()) %>% 
  ungroup() %>% 
  mutate(my_text = str_c(Classement,"\n(",round(AUC,2),")"))

#PLOT OLD SCHOOL
AUPR_data <- AUPR_data%>%
  unite(Files,Exp,cell_line)

p4_data <- AUPR_data %>% 
  filter(Files != "GSE76688_HaCaT") %>% 
  mutate(Tool = fct_reorder(Tool,AUC,mean))

cctext <- p4_data %>% filter(Files != "GSE76688_HaCaT") %>% group_by(Tool) %>% summarise(meanscore = mean(AUC),AUC = sum(AUC))

p4_data <- p4_data %>% mutate(my_text = ifelse(Tool %in% c("DeepG4","DeepG4\n(normBG)","DeepG4*"),my_text,NA))


p4_data <- p4_data %>% 
  mutate(Files =factor(Files,levels = rev(c("GSE76688_HaCaT","GSE99205_HaCaT","GSE107690_K562","GSE76688_HEKnp","GSE133379_293T","GSE133379_A549","GSE133379_H1975","GSE133379_HeLaS3")))) %>%
  mutate(col_Files = fct_recode(Files,`#2c3e50`="GSE76688_HaCaT",`#2980b9`="GSE99205_HaCaT",`#8e44ad`="GSE107690_K562",`#27ae60`="GSE76688_HEKnp",`#c0392b`="GSE133379_293T",`#d35400`="GSE133379_A549",`#16a085`="GSE133379_H1975",`#60a3bc`="GSE133379_HeLaS3"))


p4_data %>% 
  ggplot(aes(x=Tool,y=AUC,fill=Files)) +
  geom_bar(alpha=0.3,stat="identity",col="black") +
  scale_y_continuous(expand=c(0,0),limits = c(0,1.7)) +
  coord_flip(clip="off") +
  geom_text(aes(label=my_text), position=position_stack(0.45), size=4,lineheight = .8) +
  ylab("Ranking (AUPR) on 9 datasets\n(Rectangle width is proportional to AUPR)") + xlab("Tools") +
  geom_label(data=cctext,aes(x=Tool,y=AUC,label = round(meanscore,3)),fill="white",position = position_dodge(width = 1),hjust = -0.1,size=5) +
  scale_fill_manual("Dataset",values = levels(p4_data$col_Files)) -> p4
print(p4)

#PROMOTER AUPR TEST 

AUPR_data <- c(list.files("results/AUC_prom/",pattern=".tsv$",full.names=T)) %>% setNames(basename(.)) %>% map(read_tsv) %>% bind_rows(.id="file") %>% 
  filter(!str_detect(file,"qG4-ChIP-seq-of-breast-cancer-PDTX"))%>% 
  filter(!str_detect(file,"DENOISED_"))

AUPR_data <- AUPR_data %>% 
  mutate(Exp = str_extract(file,"GSE[0-9]+|breast-cancer-PDTX"))  %>% 
  mutate(Tool = str_extract(basename(file),"\\..+")) %>%
  mutate(Tool = str_remove_all(Tool,"\\.|tsv$")) %>%
  mutate(cell_line = str_extract(file,"qG4|HaCaT|K562|HEKnp|H1975|A549|293T|HeLaS3")) %>% 
  mutate(TypeExp = str_extract(file,"Peaks_G4P_G4seq|TestSet_Peaks_qG4|TestSet_Peaks_G4seqpm_BG4|TestSet_Peaks_BG4_G4seq|Promoters_qG4|Promoters_G4seq_BG4|Promoters_BG4_G4seq|Peaks_G4seqpm_BG4|Peaks_BG4_G4seq|Peaks_G4seq_qG4|Peaks_G4seq_BG4|Peaks_qG4|Peaks_G4seqpm_qG4")) 
##Classement

AUPR_data <- AUPR_data %>%
  mutate(x_axis = glue("**{cell_line}** *({Exp})*")) %>% 
  mutate(x_order = str_c(Exp,cell_line,sep="_")) %>% 
  mutate(x_order = as.numeric(as.character(fct_recode(as.factor(x_order),`1`="GSE76688_HaCaT",`2`="GSE99205_HaCaT",`3`="GSE76688_HEKnp",`4`="GSE107690_K562",`5`="GSE133379_HeLaS3",`6`="GSE133379_293T",`7`="GSE133379_A549",`8`="GSE133379_H1975")))) %>% 
  # filter(Tool %in% c("DeepG4ATACnormBG","DeepG4_classictuningOH5","G4detector_tsv","G4detector_retrained_tsv","penguinn_tsv","penguinn_retrained_tsv","quadron_retrained","quadron_score")) %>%
  filter(Tool %in% c("ATACDeepG4_ATACnormBG","DeepG4_classictuningOH5","G4detector_tsv","penguinn_tsv","quadron_score")) %>%
  # mutate(Tool_color = fct_recode(as.factor(Tool),`#e84118`="ATACDeepG4_ATACtuningOH5",`#273c75`="penguinn_retrained_tsv",`#487eb0`="penguinn_tsv",`#353b48`="G4detector_retrained_tsv",`#7f8fa6`="G4detector_tsv")) %>% 
  # mutate(Tool = fct_recode(as.factor(Tool),`DeepG4`="ATACDeepG4_ATACtuningOH5",`DeepG4*`="DeepG4_classictuningOH5",`DeepG4\n(normBG)`="DeepG4ATACnormBG",`Penguinn retrained`="penguinn_retrained_tsv",`Penguinn`="penguinn_tsv",`G4detector retrained`="G4detector_retrained_tsv",G4detector="G4detector_tsv")) %>% 
  # mutate(Tool = fct_recode(as.factor(Tool),`DeepG4*`="DeepG4_classictuningOH5",`DeepG4`="DeepG4ATACnormBG",`Penguinn retrained`="penguinn_retrained_tsv",`Penguinn`="penguinn_tsv",`G4detector retrained`="G4detector_retrained_tsv",G4detector="G4detector_tsv",`Quadron retrained`="quadron_retrained",Quadron="quadron_score")) %>% 
  mutate(Tool = fct_recode(as.factor(Tool),`DeepG4*`="DeepG4_classictuningOH5",`DeepG4`="ATACDeepG4_ATACnormBG",`Penguinn`="penguinn_tsv",G4detector="G4detector_tsv",Quadron="quadron_score")) %>% 
  # mutate(Tool = glue("<i style='color:{Tool_color}'><b>{Tool}</b></i>")) %>% 
  group_by(cell_line,Exp) %>% 
  arrange(desc(AUC)) %>% mutate(Classement = 1:dplyr::n()) %>% 
  ungroup() %>% 
  mutate(my_text = ifelse(Tool %in% c("DeepG4*","DeepG4"),str_c(Classement,"\n(",round(AUC,2),")"),
                          str_c(Classement,"\n(",round(AUC,1),")")))

AUPR_data <- AUPR_data%>%
  unite(Files,Exp,cell_line)

p4_data_prom <- AUPR_data %>% 
  filter(Files != "GSE76688_HaCaT") %>% 
  mutate(Tool = fct_reorder(Tool,AUC,mean))

cctext <- p4_data_prom %>% filter(Files != "GSE76688_HaCaT") %>% group_by(Tool) %>% summarise(meanscore = mean(AUC),AUC = sum(AUC))

# p4_data <- p4_data %>% mutate(my_text = ifelse(Tool %in% c("DeepG4","DeepG4\n(normBG)"),my_text,NA))


p4_data_prom <- p4_data_prom %>% 
  mutate(Files =factor(Files,levels = rev(c("GSE76688_HaCaT","GSE99205_HaCaT","GSE107690_K562","GSE76688_HEKnp","GSE133379_293T","GSE133379_A549","GSE133379_H1975","GSE133379_HeLaS3")))) %>%
  mutate(col_Files = fct_recode(Files,`#2c3e50`="GSE76688_HaCaT",`#2980b9`="GSE99205_HaCaT",`#8e44ad`="GSE107690_K562",`#27ae60`="GSE76688_HEKnp",`#c0392b`="GSE133379_293T",`#d35400`="GSE133379_A549",`#16a085`="GSE133379_H1975",`#60a3bc`="GSE133379_HeLaS3"))


p4_data_prom %>% 
  ggplot(aes(x=Tool,y=AUC,fill=Files)) +
  geom_bar(alpha=0.3,stat="identity",col="black") +
  scale_y_continuous(expand=c(0,0),limits = c(0,2.3)) +
  coord_flip(clip="off") +
  geom_text(aes(label=my_text), position=position_stack(0.45), size=4,lineheight = .8) +
  ylab("Ranking (AUPR) on 9 datasets\n(Rectangle width is proportional to AUPR)") + xlab("Tools") +
  geom_label(data=cctext,aes(x=Tool,y=AUC,label = round(meanscore,3)),fill="white",position = position_dodge(width = 1),hjust = -0.1,size=5) +
  scale_fill_manual("Dataset",values = levels(p4_data_prom$col_Files)) +theme(legend.position = "none") -> p4_prom
print(p4_prom)

#OTHER METRICS



AUPR_data <- c(list.files("results/other_metrics/",pattern=".tsv$",full.names=T)) %>% setNames(basename(.)) %>% map(read_tsv) %>% bind_rows(.id="file") %>% 
  filter(!str_detect(file,"qG4-ChIP-seq-of-breast-cancer-PDTX")) %>% 
  filter(!str_detect(file,"DENOISED_"))

AUPR_data <- AUPR_data %>% 
  mutate(Exp = str_extract(file,"GSE[0-9]+|breast-cancer-PDTX"))  %>% 
  mutate(Tool = str_extract(basename(file),"\\..+")) %>%
  mutate(Tool = str_remove_all(Tool,"\\.|tsv$")) %>%
  mutate(cell_line = str_extract(file,"qG4|HaCaT|K562|HEKnp|H1975|A549|293T|HeLaS3")) %>% 
  mutate(TypeExp = str_extract(file,"Peaks_G4P_G4seq|TestSet_Peaks_qG4|TestSet_Peaks_G4seqpm_BG4|TestSet_Peaks_BG4_G4seq|Promoters_qG4|Promoters_G4seq_BG4|Promoters_BG4_G4seq|Peaks_G4seqpm_BG4|Peaks_BG4_G4seq|Peaks_G4seq_qG4|Peaks_G4seq_BG4|Peaks_qG4|Peaks_G4seqpm_qG4")) 
##Classement

AUPR_data <- AUPR_data %>%
  mutate(x_axis = glue("**{cell_line}** *({Exp})*")) %>% 
  mutate(x_order = str_c(Exp,cell_line,sep="_")) %>% 
  mutate(x_order = as.numeric(as.character(fct_recode(as.factor(x_order),`1`="GSE76688_HaCaT",`2`="GSE99205_HaCaT",`3`="GSE76688_HEKnp",`4`="GSE107690_K562",`5`="GSE133379_HeLaS3",`6`="GSE133379_293T",`7`="GSE133379_A549",`8`="GSE133379_H1975")))) %>% 
  # filter(Tool %in% c("DeepG4ATACnormBG","DeepG4_classictuningOH5","G4detector_tsv","G4detector_retrained_tsv","penguinn_tsv","penguinn_retrained_tsv","quadron_retrained","quadron_score")) %>%
  filter(Tool %in% c("ATACDeepG4_ATACnormBG","DeepG4_classictuningOH5","G4detector_tsv","penguinn_tsv")) %>%
  # mutate(Tool_color = fct_recode(as.factor(Tool),`#e84118`="ATACDeepG4_ATACtuningOH5",`#273c75`="penguinn_retrained_tsv",`#487eb0`="penguinn_tsv",`#353b48`="G4detector_retrained_tsv",`#7f8fa6`="G4detector_tsv")) %>% 
  # mutate(Tool = fct_recode(as.factor(Tool),`DeepG4`="ATACDeepG4_ATACtuningOH5",`DeepG4*`="DeepG4_classictuningOH5",`DeepG4\n(normBG)`="DeepG4ATACnormBG",`Penguinn retrained`="penguinn_retrained_tsv",`Penguinn`="penguinn_tsv",`G4detector retrained`="G4detector_retrained_tsv",G4detector="G4detector_tsv")) %>% 
  # mutate(Tool = fct_recode(as.factor(Tool),`DeepG4*`="DeepG4_classictuningOH5",`DeepG4`="DeepG4ATACnormBG",`Penguinn retrained`="penguinn_retrained_tsv",`Penguinn`="penguinn_tsv",`G4detector retrained`="G4detector_retrained_tsv",G4detector="G4detector_tsv",`Quadron retrained`="quadron_retrained",Quadron="quadron_score")) %>% 
  mutate(Tool = fct_recode(as.factor(Tool),`DeepG4*`="DeepG4_classictuningOH5",`DeepG4`="ATACDeepG4_ATACnormBG",`Penguinn`="penguinn_tsv",G4detector="G4detector_tsv")) %>% 
  # mutate(Tool = glue("<i style='color:{Tool_color}'><b>{Tool}</b></i>")) %>% 
  group_by(cell_line,Exp) %>% 
  arrange(desc(accuracy)) %>% mutate(Classement = 1:dplyr::n()) %>% 
  ungroup() %>% 
  mutate(my_text = str_c(Classement,"\n(",round(accuracy,3),")"))

#PLOT OLD SCHOOL
AUPR_data <- AUPR_data%>%
  unite(Files,Exp,cell_line)

p4_data <- AUPR_data %>% 
  filter(Files != "GSE76688_HaCaT") %>% 
  mutate(Tool = fct_reorder(Tool,accuracy,mean))

cctext <- p4_data %>% filter(Files != "GSE76688_HaCaT") %>% group_by(Tool) %>% summarise(meanscore = mean(accuracy),accuracy = sum(accuracy))

# p4_data <- p4_data %>% mutate(my_text = ifelse(Tool %in% c("DeepG4","DeepG4\n(normBG)","DeepG4*"),my_text,NA))


p4_data <- p4_data %>% 
  mutate(Files =factor(Files,levels = rev(c("GSE76688_HaCaT","GSE99205_HaCaT","GSE107690_K562","GSE76688_HEKnp","GSE133379_293T","GSE133379_A549","GSE133379_H1975","GSE133379_HeLaS3")))) %>%
  mutate(col_Files = fct_recode(Files,`#2c3e50`="GSE76688_HaCaT",`#2980b9`="GSE99205_HaCaT",`#8e44ad`="GSE107690_K562",`#27ae60`="GSE76688_HEKnp",`#c0392b`="GSE133379_293T",`#d35400`="GSE133379_A549",`#16a085`="GSE133379_H1975",`#60a3bc`="GSE133379_HeLaS3"))


p4_data %>% 
  ggplot(aes(x=Tool,y=accuracy,fill=Files)) +
  geom_bar(alpha=0.3,stat="identity",col="black") +
  scale_y_continuous(expand=c(0,0),limits = c(0,7.5)) +
  coord_flip(clip="off") +
  geom_text(aes(label=my_text), position=position_stack(0.45), size=4,lineheight = .8) +
  ylab("Ranking (Accuracy) on 9 datasets\n(Rectangle width is proportional to Accuracy)") + xlab("Tools") +
  geom_label(data=cctext,aes(x=Tool,y=accuracy,label = round(meanscore,3)),fill="white",position = position_dodge(width = 1),hjust = -0.1,size=5) +
  scale_fill_manual("Dataset",values = levels(p4_data$col_Files)) + theme(axis.text.y = element_text(color = "black",size=16)) -> p4_acc
print(p4_acc)

#PROMOTER AUPR TEST 

AUPR_data <- c(list.files("results/other_metrics_prom/",pattern=".tsv$",full.names=T)) %>% setNames(basename(.)) %>% map(read_tsv) %>% bind_rows(.id="file") %>% 
  filter(!str_detect(file,"qG4-ChIP-seq-of-breast-cancer-PDTX"))%>% 
  filter(!str_detect(file,"DENOISED_"))

AUPR_data <- AUPR_data %>% 
  mutate(Exp = str_extract(file,"GSE[0-9]+|breast-cancer-PDTX"))  %>% 
  mutate(Tool = str_extract(basename(file),"\\..+")) %>%
  mutate(Tool = str_remove_all(Tool,"\\.|tsv$")) %>%
  mutate(cell_line = str_extract(file,"qG4|HaCaT|K562|HEKnp|H1975|A549|293T|HeLaS3")) %>% 
  mutate(TypeExp = str_extract(file,"Peaks_G4P_G4seq|TestSet_Peaks_qG4|TestSet_Peaks_G4seqpm_BG4|TestSet_Peaks_BG4_G4seq|Promoters_qG4|Promoters_G4seq_BG4|Promoters_BG4_G4seq|Peaks_G4seqpm_BG4|Peaks_BG4_G4seq|Peaks_G4seq_qG4|Peaks_G4seq_BG4|Peaks_qG4|Peaks_G4seqpm_qG4")) 
##Classement

AUPR_data <- AUPR_data %>%
  mutate(x_axis = glue("**{cell_line}** *({Exp})*")) %>% 
  mutate(x_order = str_c(Exp,cell_line,sep="_")) %>% 
  mutate(x_order = as.numeric(as.character(fct_recode(as.factor(x_order),`1`="GSE76688_HaCaT",`2`="GSE99205_HaCaT",`3`="GSE76688_HEKnp",`4`="GSE107690_K562",`5`="GSE133379_HeLaS3",`6`="GSE133379_293T",`7`="GSE133379_A549",`8`="GSE133379_H1975")))) %>% 
  # filter(Tool %in% c("DeepG4ATACnormBG","DeepG4_classictuningOH5","G4detector_tsv","G4detector_retrained_tsv","penguinn_tsv","penguinn_retrained_tsv","quadron_retrained","quadron_score")) %>%
  filter(Tool %in% c("ATACDeepG4_ATACnormBG","DeepG4_classictuningOH5","G4detector_tsv","penguinn_tsv")) %>%
  # mutate(Tool_color = fct_recode(as.factor(Tool),`#e84118`="ATACDeepG4_ATACtuningOH5",`#273c75`="penguinn_retrained_tsv",`#487eb0`="penguinn_tsv",`#353b48`="G4detector_retrained_tsv",`#7f8fa6`="G4detector_tsv")) %>% 
  # mutate(Tool = fct_recode(as.factor(Tool),`DeepG4`="ATACDeepG4_ATACtuningOH5",`DeepG4*`="DeepG4_classictuningOH5",`DeepG4\n(normBG)`="DeepG4ATACnormBG",`Penguinn retrained`="penguinn_retrained_tsv",`Penguinn`="penguinn_tsv",`G4detector retrained`="G4detector_retrained_tsv",G4detector="G4detector_tsv")) %>% 
  # mutate(Tool = fct_recode(as.factor(Tool),`DeepG4*`="DeepG4_classictuningOH5",`DeepG4`="DeepG4ATACnormBG",`Penguinn retrained`="penguinn_retrained_tsv",`Penguinn`="penguinn_tsv",`G4detector retrained`="G4detector_retrained_tsv",G4detector="G4detector_tsv",`Quadron retrained`="quadron_retrained",Quadron="quadron_score")) %>% 
  mutate(Tool = fct_recode(as.factor(Tool),`DeepG4*`="DeepG4_classictuningOH5",`DeepG4`="ATACDeepG4_ATACnormBG",`Penguinn`="penguinn_tsv",G4detector="G4detector_tsv")) %>% 
  # mutate(Tool = glue("<i style='color:{Tool_color}'><b>{Tool}</b></i>")) %>% 
  group_by(cell_line,Exp) %>% 
  arrange(desc(accuracy)) %>% mutate(Classement = 1:dplyr::n()) %>% 
  ungroup() %>% 
  mutate(my_text = ifelse(Tool %in% c("DeepG4*","DeepG4"),str_c(Classement,"\n(",round(accuracy,2),")"),
                          str_c(Classement,"\n(",round(accuracy,1),")")))

AUPR_data <- AUPR_data%>%
  unite(Files,Exp,cell_line)

p4_data_prom <- AUPR_data %>% 
  filter(Files != "GSE76688_HaCaT") %>% 
  mutate(Tool = fct_reorder(Tool,accuracy,mean))

cctext <- p4_data_prom %>% filter(Files != "GSE76688_HaCaT") %>% group_by(Tool) %>% summarise(meanscore = mean(accuracy),accuracy = sum(accuracy))

# p4_data <- p4_data %>% mutate(my_text = ifelse(Tool %in% c("DeepG4","DeepG4\n(normBG)"),my_text,NA))


p4_data_prom <- p4_data_prom %>% 
  mutate(Files =factor(Files,levels = rev(c("GSE76688_HaCaT","GSE99205_HaCaT","GSE107690_K562","GSE76688_HEKnp","GSE133379_293T","GSE133379_A549","GSE133379_H1975","GSE133379_HeLaS3")))) %>%
  mutate(col_Files = fct_recode(Files,`#2c3e50`="GSE76688_HaCaT",`#2980b9`="GSE99205_HaCaT",`#8e44ad`="GSE107690_K562",`#27ae60`="GSE76688_HEKnp",`#c0392b`="GSE133379_293T",`#d35400`="GSE133379_A549",`#16a085`="GSE133379_H1975",`#60a3bc`="GSE133379_HeLaS3"))


p4_data_prom %>% 
  ggplot(aes(x=Tool,y=accuracy,fill=Files)) +
  geom_bar(alpha=0.3,stat="identity",col="black") +
  scale_y_continuous(expand=c(0,0),limits = c(0,5.5)) +
  coord_flip(clip="off") +
  geom_text(aes(label=my_text), position=position_stack(0.45), size=4,lineheight = .8) +
  ylab("Ranking (Accuracy) on 9 datasets\n(Rectangle width is proportional to Accuracy)") + xlab("Tools") +
  geom_label(data=cctext,aes(x=Tool,y=accuracy,label = round(meanscore,3)),fill="white",position = position_dodge(width = 1),hjust = -0.1,size=5) +
  scale_fill_manual("Dataset",values = levels(p4_data_prom$col_Files)) +theme(legend.position = "none",axis.text.y = element_text(color = "black",size=16)) -> p4_acc_prom
print(p4_acc_prom)

#1-FDR



AUPR_data <- c(list.files("results/other_metrics/",pattern=".tsv$",full.names=T)) %>% setNames(basename(.)) %>% map(read_tsv) %>% bind_rows(.id="file") %>% 
  filter(!str_detect(file,"qG4-ChIP-seq-of-breast-cancer-PDTX")) %>% 
  filter(!str_detect(file,"DENOISED_"))

AUPR_data <- AUPR_data %>% 
  mutate(precision = 1-FDR) %>% dplyr::select(-FDR) %>% 
  mutate(Exp = str_extract(file,"GSE[0-9]+|breast-cancer-PDTX"))  %>% 
  mutate(Tool = str_extract(basename(file),"\\..+")) %>%
  mutate(Tool = str_remove_all(Tool,"\\.|tsv$")) %>%
  mutate(cell_line = str_extract(file,"qG4|HaCaT|K562|HEKnp|H1975|A549|293T|HeLaS3")) %>% 
  mutate(TypeExp = str_extract(file,"Peaks_G4P_G4seq|TestSet_Peaks_qG4|TestSet_Peaks_G4seqpm_BG4|TestSet_Peaks_BG4_G4seq|Promoters_qG4|Promoters_G4seq_BG4|Promoters_BG4_G4seq|Peaks_G4seqpm_BG4|Peaks_BG4_G4seq|Peaks_G4seq_qG4|Peaks_G4seq_BG4|Peaks_qG4|Peaks_G4seqpm_qG4")) 
##Classement

AUPR_data <- AUPR_data %>%
  mutate(x_axis = glue("**{cell_line}** *({Exp})*")) %>% 
  mutate(x_order = str_c(Exp,cell_line,sep="_")) %>% 
  mutate(x_order = as.numeric(as.character(fct_recode(as.factor(x_order),`1`="GSE76688_HaCaT",`2`="GSE99205_HaCaT",`3`="GSE76688_HEKnp",`4`="GSE107690_K562",`5`="GSE133379_HeLaS3",`6`="GSE133379_293T",`7`="GSE133379_A549",`8`="GSE133379_H1975")))) %>% 
  # filter(Tool %in% c("DeepG4ATACnormBG","DeepG4_classictuningOH5","G4detector_tsv","G4detector_retrained_tsv","penguinn_tsv","penguinn_retrained_tsv","quadron_retrained","quadron_score")) %>%
  filter(Tool %in% c("ATACDeepG4_ATACnormBG","DeepG4_classictuningOH5","G4detector_tsv","penguinn_tsv")) %>%
  # mutate(Tool_color = fct_recode(as.factor(Tool),`#e84118`="ATACDeepG4_ATACtuningOH5",`#273c75`="penguinn_retrained_tsv",`#487eb0`="penguinn_tsv",`#353b48`="G4detector_retrained_tsv",`#7f8fa6`="G4detector_tsv")) %>% 
  # mutate(Tool = fct_recode(as.factor(Tool),`DeepG4`="ATACDeepG4_ATACtuningOH5",`DeepG4*`="DeepG4_classictuningOH5",`DeepG4\n(normBG)`="DeepG4ATACnormBG",`Penguinn retrained`="penguinn_retrained_tsv",`Penguinn`="penguinn_tsv",`G4detector retrained`="G4detector_retrained_tsv",G4detector="G4detector_tsv")) %>% 
  # mutate(Tool = fct_recode(as.factor(Tool),`DeepG4*`="DeepG4_classictuningOH5",`DeepG4`="DeepG4ATACnormBG",`Penguinn retrained`="penguinn_retrained_tsv",`Penguinn`="penguinn_tsv",`G4detector retrained`="G4detector_retrained_tsv",G4detector="G4detector_tsv",`Quadron retrained`="quadron_retrained",Quadron="quadron_score")) %>% 
  mutate(Tool = fct_recode(as.factor(Tool),`DeepG4*`="DeepG4_classictuningOH5",`DeepG4`="ATACDeepG4_ATACnormBG",`Penguinn`="penguinn_tsv",G4detector="G4detector_tsv")) %>% 
  # mutate(Tool = glue("<i style='color:{Tool_color}'><b>{Tool}</b></i>")) %>% 
  group_by(cell_line,Exp) %>% 
  arrange(desc(precision)) %>% mutate(Classement = 1:dplyr::n()) %>% 
  ungroup() %>% 
  mutate(my_text = str_c(Classement,"\n(",round(precision,3),")"))

#PLOT OLD SCHOOL
AUPR_data <- AUPR_data%>%
  unite(Files,Exp,cell_line)

p4_data <- AUPR_data %>% 
  filter(Files != "GSE76688_HaCaT") %>% 
  mutate(Tool = fct_reorder(Tool,precision,mean))

cctext <- p4_data %>% filter(Files != "GSE76688_HaCaT") %>% group_by(Tool) %>% summarise(meanscore = mean(precision),precision = sum(precision))

# p4_data <- p4_data %>% mutate(my_text = ifelse(Tool %in% c("DeepG4","DeepG4\n(normBG)","DeepG4*"),my_text,NA))


p4_data <- p4_data %>% 
  mutate(Files =factor(Files,levels = rev(c("GSE76688_HaCaT","GSE99205_HaCaT","GSE107690_K562","GSE76688_HEKnp","GSE133379_293T","GSE133379_A549","GSE133379_H1975","GSE133379_HeLaS3")))) %>%
  mutate(col_Files = fct_recode(Files,`#2c3e50`="GSE76688_HaCaT",`#2980b9`="GSE99205_HaCaT",`#8e44ad`="GSE107690_K562",`#27ae60`="GSE76688_HEKnp",`#c0392b`="GSE133379_293T",`#d35400`="GSE133379_A549",`#16a085`="GSE133379_H1975",`#60a3bc`="GSE133379_HeLaS3"))


p4_data %>% 
  ggplot(aes(x=Tool,y=precision,fill=Files)) +
  geom_bar(alpha=0.3,stat="identity",col="black") +
  scale_y_continuous(expand=c(0,0),limits = c(0,0.8)) +
  coord_flip(clip="off") +
  geom_text(aes(label=my_text), position=position_stack(0.45), size=4,lineheight = .8) +
  ylab("Ranking (Precision) on 9 datasets\n(Rectangle width is proportional to Precision)") + xlab("Tools") +
  geom_label(data=cctext,aes(x=Tool,y=precision,label = round(meanscore,3)),fill="white",position = position_dodge(width = 1),hjust = -0.1,size=5) +
  scale_fill_manual("Dataset",values = levels(p4_data$col_Files)) +theme(axis.text.y = element_text(color = "black",size=16)) -> p4_fdr
print(p4_fdr)

#PROMOTER AUPR TEST 

AUPR_data <- c(list.files("results/other_metrics_prom/",pattern=".tsv$",full.names=T)) %>% setNames(basename(.)) %>% map(read_tsv) %>% bind_rows(.id="file") %>% 
  filter(!str_detect(file,"qG4-ChIP-seq-of-breast-cancer-PDTX"))%>% 
  filter(!str_detect(file,"DENOISED_"))

AUPR_data <- AUPR_data %>% 
  mutate(precision = 1-FDR) %>% dplyr::select(-FDR) %>% 
  mutate(Exp = str_extract(file,"GSE[0-9]+|breast-cancer-PDTX"))  %>% 
  mutate(Tool = str_extract(basename(file),"\\..+")) %>%
  mutate(Tool = str_remove_all(Tool,"\\.|tsv$")) %>%
  mutate(cell_line = str_extract(file,"qG4|HaCaT|K562|HEKnp|H1975|A549|293T|HeLaS3")) %>% 
  mutate(TypeExp = str_extract(file,"Peaks_G4P_G4seq|TestSet_Peaks_qG4|TestSet_Peaks_G4seqpm_BG4|TestSet_Peaks_BG4_G4seq|Promoters_qG4|Promoters_G4seq_BG4|Promoters_BG4_G4seq|Peaks_G4seqpm_BG4|Peaks_BG4_G4seq|Peaks_G4seq_qG4|Peaks_G4seq_BG4|Peaks_qG4|Peaks_G4seqpm_qG4")) 
##Classement

AUPR_data <- AUPR_data %>%
  mutate(x_axis = glue("**{cell_line}** *({Exp})*")) %>% 
  mutate(x_order = str_c(Exp,cell_line,sep="_")) %>% 
  mutate(x_order = as.numeric(as.character(fct_recode(as.factor(x_order),`1`="GSE76688_HaCaT",`2`="GSE99205_HaCaT",`3`="GSE76688_HEKnp",`4`="GSE107690_K562",`5`="GSE133379_HeLaS3",`6`="GSE133379_293T",`7`="GSE133379_A549",`8`="GSE133379_H1975")))) %>% 
  # filter(Tool %in% c("DeepG4ATACnormBG","DeepG4_classictuningOH5","G4detector_tsv","G4detector_retrained_tsv","penguinn_tsv","penguinn_retrained_tsv","quadron_retrained","quadron_score")) %>%
  filter(Tool %in% c("ATACDeepG4_ATACnormBG","DeepG4_classictuningOH5","G4detector_tsv","penguinn_tsv")) %>%
  # mutate(Tool_color = fct_recode(as.factor(Tool),`#e84118`="ATACDeepG4_ATACtuningOH5",`#273c75`="penguinn_retrained_tsv",`#487eb0`="penguinn_tsv",`#353b48`="G4detector_retrained_tsv",`#7f8fa6`="G4detector_tsv")) %>% 
  # mutate(Tool = fct_recode(as.factor(Tool),`DeepG4`="ATACDeepG4_ATACtuningOH5",`DeepG4*`="DeepG4_classictuningOH5",`DeepG4\n(normBG)`="DeepG4ATACnormBG",`Penguinn retrained`="penguinn_retrained_tsv",`Penguinn`="penguinn_tsv",`G4detector retrained`="G4detector_retrained_tsv",G4detector="G4detector_tsv")) %>% 
  # mutate(Tool = fct_recode(as.factor(Tool),`DeepG4*`="DeepG4_classictuningOH5",`DeepG4`="DeepG4ATACnormBG",`Penguinn retrained`="penguinn_retrained_tsv",`Penguinn`="penguinn_tsv",`G4detector retrained`="G4detector_retrained_tsv",G4detector="G4detector_tsv",`Quadron retrained`="quadron_retrained",Quadron="quadron_score")) %>% 
  mutate(Tool = fct_recode(as.factor(Tool),`DeepG4*`="DeepG4_classictuningOH5",`DeepG4`="ATACDeepG4_ATACnormBG",`Penguinn`="penguinn_tsv",G4detector="G4detector_tsv")) %>% 
  # mutate(Tool = glue("<i style='color:{Tool_color}'><b>{Tool}</b></i>")) %>% 
  group_by(cell_line,Exp) %>% 
  arrange(desc(precision)) %>% mutate(Classement = 1:dplyr::n()) %>% 
  ungroup() %>% 
  mutate(my_text = ifelse(Tool %in% c("DeepG4*","DeepG4"),str_c(Classement,"\n(",round(precision,2),")"),
                          str_c(Classement,"\n(",round(precision,1),")")))

AUPR_data <- AUPR_data%>%
  unite(Files,Exp,cell_line)

p4_data_prom <- AUPR_data %>% 
  filter(Files != "GSE76688_HaCaT") %>% 
  mutate(Tool = fct_reorder(Tool,precision,mean))

cctext <- p4_data_prom %>% filter(Files != "GSE76688_HaCaT") %>% group_by(Tool) %>% summarise(meanscore = mean(precision),precision = sum(precision))

# p4_data <- p4_data %>% mutate(my_text = ifelse(Tool %in% c("DeepG4","DeepG4\n(normBG)"),my_text,NA))


p4_data_prom <- p4_data_prom %>% 
  mutate(Files =factor(Files,levels = rev(c("GSE76688_HaCaT","GSE99205_HaCaT","GSE107690_K562","GSE76688_HEKnp","GSE133379_293T","GSE133379_A549","GSE133379_H1975","GSE133379_HeLaS3")))) %>%
  mutate(col_Files = fct_recode(Files,`#2c3e50`="GSE76688_HaCaT",`#2980b9`="GSE99205_HaCaT",`#8e44ad`="GSE107690_K562",`#27ae60`="GSE76688_HEKnp",`#c0392b`="GSE133379_293T",`#d35400`="GSE133379_A549",`#16a085`="GSE133379_H1975",`#60a3bc`="GSE133379_HeLaS3"))


p4_data_prom %>% 
  ggplot(aes(x=Tool,y=precision,fill=Files)) +
  geom_bar(alpha=0.3,stat="identity",col="black") +
  scale_y_continuous(expand=c(0,0),limits = c(0,1.5)) +
  coord_flip(clip="off") +
  geom_text(aes(label=my_text), position=position_stack(0.45), size=4,lineheight = .8) +
  ylab("Ranking (Precision) on 9 datasets\n(Rectangle width is proportional to Precision)") + xlab("Tools") +
  geom_label(data=cctext,aes(x=Tool,y=precision,label = round(meanscore,3)),fill="white",position = position_dodge(width = 1),hjust = -0.1,size=5) +
  scale_fill_manual("Dataset",values = levels(p4_data_prom$col_Files)) +theme(legend.position = "none",axis.text.y = element_text(color = "black",size=16)) -> p4_fdr_prom
print(p4_fdr_prom)

#FDR




AUPR_data <- c(list.files("results/other_metrics/",pattern=".tsv$",full.names=T)) %>% setNames(basename(.)) %>% map(read_tsv) %>% bind_rows(.id="file") %>% 
  filter(!str_detect(file,"qG4-ChIP-seq-of-breast-cancer-PDTX")) %>% 
  filter(!str_detect(file,"DENOISED_"))

AUPR_data <- AUPR_data %>% 
  mutate(Exp = str_extract(file,"GSE[0-9]+|breast-cancer-PDTX"))  %>% 
  mutate(Tool = str_extract(basename(file),"\\..+")) %>%
  mutate(Tool = str_remove_all(Tool,"\\.|tsv$")) %>%
  mutate(cell_line = str_extract(file,"qG4|HaCaT|K562|HEKnp|H1975|A549|293T|HeLaS3")) %>% 
  mutate(TypeExp = str_extract(file,"Peaks_G4P_G4seq|TestSet_Peaks_qG4|TestSet_Peaks_G4seqpm_BG4|TestSet_Peaks_BG4_G4seq|Promoters_qG4|Promoters_G4seq_BG4|Promoters_BG4_G4seq|Peaks_G4seqpm_BG4|Peaks_BG4_G4seq|Peaks_G4seq_qG4|Peaks_G4seq_BG4|Peaks_qG4|Peaks_G4seqpm_qG4")) 
##Classement

AUPR_data <- AUPR_data %>%
  mutate(x_axis = glue("**{cell_line}** *({Exp})*")) %>% 
  mutate(x_order = str_c(Exp,cell_line,sep="_")) %>% 
  mutate(x_order = as.numeric(as.character(fct_recode(as.factor(x_order),`1`="GSE76688_HaCaT",`2`="GSE99205_HaCaT",`3`="GSE76688_HEKnp",`4`="GSE107690_K562",`5`="GSE133379_HeLaS3",`6`="GSE133379_293T",`7`="GSE133379_A549",`8`="GSE133379_H1975")))) %>% 
  # filter(Tool %in% c("DeepG4ATACnormBG","DeepG4_classictuningOH5","G4detector_tsv","G4detector_retrained_tsv","penguinn_tsv","penguinn_retrained_tsv","quadron_retrained","quadron_score")) %>%
  filter(Tool %in% c("ATACDeepG4_ATACnormBG","DeepG4_classictuningOH5","G4detector_tsv","penguinn_tsv")) %>%
  # mutate(Tool_color = fct_recode(as.factor(Tool),`#e84118`="ATACDeepG4_ATACtuningOH5",`#273c75`="penguinn_retrained_tsv",`#487eb0`="penguinn_tsv",`#353b48`="G4detector_retrained_tsv",`#7f8fa6`="G4detector_tsv")) %>% 
  # mutate(Tool = fct_recode(as.factor(Tool),`DeepG4`="ATACDeepG4_ATACtuningOH5",`DeepG4*`="DeepG4_classictuningOH5",`DeepG4\n(normBG)`="DeepG4ATACnormBG",`Penguinn retrained`="penguinn_retrained_tsv",`Penguinn`="penguinn_tsv",`G4detector retrained`="G4detector_retrained_tsv",G4detector="G4detector_tsv")) %>% 
  # mutate(Tool = fct_recode(as.factor(Tool),`DeepG4*`="DeepG4_classictuningOH5",`DeepG4`="DeepG4ATACnormBG",`Penguinn retrained`="penguinn_retrained_tsv",`Penguinn`="penguinn_tsv",`G4detector retrained`="G4detector_retrained_tsv",G4detector="G4detector_tsv",`Quadron retrained`="quadron_retrained",Quadron="quadron_score")) %>% 
  mutate(Tool = fct_recode(as.factor(Tool),`DeepG4*`="DeepG4_classictuningOH5",`DeepG4`="ATACDeepG4_ATACnormBG",`Penguinn`="penguinn_tsv",G4detector="G4detector_tsv")) %>% 
  # mutate(Tool = glue("<i style='color:{Tool_color}'><b>{Tool}</b></i>")) %>% 
  group_by(cell_line,Exp) %>% 
  arrange(desc(FDR)) %>% mutate(Classement = 1:dplyr::n()) %>% 
  ungroup() %>% 
  mutate(my_text = str_c(Classement,"\n(",round(FDR,3),")"))

#PLOT OLD SCHOOL
AUPR_data <- AUPR_data%>%
  unite(Files,Exp,cell_line)

p4_data <- AUPR_data %>% 
  filter(Files != "GSE76688_HaCaT") %>% 
  mutate(Tool = fct_reorder(Tool,FDR,mean))

cctext <- p4_data %>% filter(Files != "GSE76688_HaCaT") %>% group_by(Tool) %>% summarise(meanscore = mean(FDR),FDR = sum(FDR))

# p4_data <- p4_data %>% mutate(my_text = ifelse(Tool %in% c("DeepG4","DeepG4\n(normBG)","DeepG4*"),my_text,NA))


p4_data <- p4_data %>% 
  mutate(Files =factor(Files,levels = rev(c("GSE76688_HaCaT","GSE99205_HaCaT","GSE107690_K562","GSE76688_HEKnp","GSE133379_293T","GSE133379_A549","GSE133379_H1975","GSE133379_HeLaS3")))) %>%
  mutate(col_Files = fct_recode(Files,`#2c3e50`="GSE76688_HaCaT",`#2980b9`="GSE99205_HaCaT",`#8e44ad`="GSE107690_K562",`#27ae60`="GSE76688_HEKnp",`#c0392b`="GSE133379_293T",`#d35400`="GSE133379_A549",`#16a085`="GSE133379_H1975",`#60a3bc`="GSE133379_HeLaS3"))


p4_data %>% 
  ggplot(aes(x=Tool,y=FDR,fill=Files)) +
  geom_bar(alpha=0.3,stat="identity",col="black") +
  scale_y_continuous(expand=c(0,0),limits = c(0,8)) +
  coord_flip(clip="off") +
  geom_text(aes(label=my_text), position=position_stack(0.45), size=4,lineheight = .8) +
  ylab("Ranking (FDR) on 9 datasets\n(Rectangle width is proportional to FDR)") + xlab("Tools") +
  geom_label(data=cctext,aes(x=Tool,y=FDR,label = round(meanscore,3)),fill="white",position = position_dodge(width = 1),hjust = -0.1,size=5) +
  scale_fill_manual("Dataset",values = levels(p4_data$col_Files)) +theme(axis.text.y = element_text(color = "black",size=16)) -> p4_fdr
print(p4_fdr)

#PROMOTER AUPR TEST 

AUPR_data <- c(list.files("results/other_metrics_prom/",pattern=".tsv$",full.names=T)) %>% setNames(basename(.)) %>% map(read_tsv) %>% bind_rows(.id="file") %>% 
  filter(!str_detect(file,"qG4-ChIP-seq-of-breast-cancer-PDTX"))%>% 
  filter(!str_detect(file,"DENOISED_"))

AUPR_data <- AUPR_data %>% 
  mutate(Exp = str_extract(file,"GSE[0-9]+|breast-cancer-PDTX"))  %>% 
  mutate(Tool = str_extract(basename(file),"\\..+")) %>%
  mutate(Tool = str_remove_all(Tool,"\\.|tsv$")) %>%
  mutate(cell_line = str_extract(file,"qG4|HaCaT|K562|HEKnp|H1975|A549|293T|HeLaS3")) %>% 
  mutate(TypeExp = str_extract(file,"Peaks_G4P_G4seq|TestSet_Peaks_qG4|TestSet_Peaks_G4seqpm_BG4|TestSet_Peaks_BG4_G4seq|Promoters_qG4|Promoters_G4seq_BG4|Promoters_BG4_G4seq|Peaks_G4seqpm_BG4|Peaks_BG4_G4seq|Peaks_G4seq_qG4|Peaks_G4seq_BG4|Peaks_qG4|Peaks_G4seqpm_qG4")) 
##Classement

AUPR_data <- AUPR_data %>%
  mutate(x_axis = glue("**{cell_line}** *({Exp})*")) %>% 
  mutate(x_order = str_c(Exp,cell_line,sep="_")) %>% 
  mutate(x_order = as.numeric(as.character(fct_recode(as.factor(x_order),`1`="GSE76688_HaCaT",`2`="GSE99205_HaCaT",`3`="GSE76688_HEKnp",`4`="GSE107690_K562",`5`="GSE133379_HeLaS3",`6`="GSE133379_293T",`7`="GSE133379_A549",`8`="GSE133379_H1975")))) %>% 
  # filter(Tool %in% c("DeepG4ATACnormBG","DeepG4_classictuningOH5","G4detector_tsv","G4detector_retrained_tsv","penguinn_tsv","penguinn_retrained_tsv","quadron_retrained","quadron_score")) %>%
  filter(Tool %in% c("ATACDeepG4_ATACnormBG","DeepG4_classictuningOH5","G4detector_tsv","penguinn_tsv")) %>%
  # mutate(Tool_color = fct_recode(as.factor(Tool),`#e84118`="ATACDeepG4_ATACtuningOH5",`#273c75`="penguinn_retrained_tsv",`#487eb0`="penguinn_tsv",`#353b48`="G4detector_retrained_tsv",`#7f8fa6`="G4detector_tsv")) %>% 
  # mutate(Tool = fct_recode(as.factor(Tool),`DeepG4`="ATACDeepG4_ATACtuningOH5",`DeepG4*`="DeepG4_classictuningOH5",`DeepG4\n(normBG)`="DeepG4ATACnormBG",`Penguinn retrained`="penguinn_retrained_tsv",`Penguinn`="penguinn_tsv",`G4detector retrained`="G4detector_retrained_tsv",G4detector="G4detector_tsv")) %>% 
  # mutate(Tool = fct_recode(as.factor(Tool),`DeepG4*`="DeepG4_classictuningOH5",`DeepG4`="DeepG4ATACnormBG",`Penguinn retrained`="penguinn_retrained_tsv",`Penguinn`="penguinn_tsv",`G4detector retrained`="G4detector_retrained_tsv",G4detector="G4detector_tsv",`Quadron retrained`="quadron_retrained",Quadron="quadron_score")) %>% 
  mutate(Tool = fct_recode(as.factor(Tool),`DeepG4*`="DeepG4_classictuningOH5",`DeepG4`="ATACDeepG4_ATACnormBG",`Penguinn`="penguinn_tsv",G4detector="G4detector_tsv")) %>% 
  # mutate(Tool = glue("<i style='color:{Tool_color}'><b>{Tool}</b></i>")) %>% 
  group_by(cell_line,Exp) %>% 
  arrange(desc(FDR)) %>% mutate(Classement = 1:dplyr::n()) %>% 
  ungroup() %>% 
  mutate(my_text = ifelse(Tool %in% c("DeepG4*","DeepG4"),str_c(Classement,"\n(",round(FDR,2),")"),
                          str_c(Classement,"\n(",round(FDR,1),")")))

AUPR_data <- AUPR_data%>%
  unite(Files,Exp,cell_line)

p4_data_prom <- AUPR_data %>% 
  filter(Files != "GSE76688_HaCaT") %>% 
  mutate(Tool = fct_reorder(Tool,FDR,mean))

cctext <- p4_data_prom %>% filter(Files != "GSE76688_HaCaT") %>% group_by(Tool) %>% summarise(meanscore = mean(FDR),FDR = sum(FDR))

# p4_data <- p4_data %>% mutate(my_text = ifelse(Tool %in% c("DeepG4","DeepG4\n(normBG)"),my_text,NA))


p4_data_prom <- p4_data_prom %>% 
  mutate(Files =factor(Files,levels = rev(c("GSE76688_HaCaT","GSE99205_HaCaT","GSE107690_K562","GSE76688_HEKnp","GSE133379_293T","GSE133379_A549","GSE133379_H1975","GSE133379_HeLaS3")))) %>%
  mutate(col_Files = fct_recode(Files,`#2c3e50`="GSE76688_HaCaT",`#2980b9`="GSE99205_HaCaT",`#8e44ad`="GSE107690_K562",`#27ae60`="GSE76688_HEKnp",`#c0392b`="GSE133379_293T",`#d35400`="GSE133379_A549",`#16a085`="GSE133379_H1975",`#60a3bc`="GSE133379_HeLaS3"))


p4_data_prom %>% 
  ggplot(aes(x=Tool,y=FDR,fill=Files)) +
  geom_bar(alpha=0.3,stat="identity",col="black") +
  scale_y_continuous(expand=c(0,0),limits = c(0,7)) +
  coord_flip(clip="off") +
  geom_text(aes(label=my_text), position=position_stack(0.45), size=4,lineheight = .8) +
  ylab("Ranking (FDR) on 9 datasets\n(Rectangle width is proportional to FDR)") + xlab("Tools") +
  geom_label(data=cctext,aes(x=Tool,y=FDR,label = round(meanscore,3)),fill="white",position = position_dodge(width = 1),hjust = -0.1,size=5) +
  scale_fill_manual("Dataset",values = levels(p4_data_prom$col_Files)) +theme(legend.position = "none",axis.text.y = element_text(color = "black",size=16)) -> p4_fdr_prom
print(p4_fdr_prom)


```